datafusion/src/scalar.rs

```diff
@@ -64,7 +70,10 @@ pub enum ScalarValue {
     /// large binary
     LargeBinary(Option<Vec<u8>>),
     /// list of nested ScalarValue
-    List(Option<Vec<ScalarValue>>, DataType),
+    // 1st argument are the inner values (e.g. Int64Array)
+    // 2st argument is the Lists' datatype (i.e. it includes `Field`)
+    // to downcast inner values, use ListArray::<i32>::get_child()
+    List(Option<Arc<dyn Array>>, DataType),
```

Comment:
@jorgecarleitao do you remember what's reason for changing from vector of ScalarValue to Arrow array here? Is it for static typing of the individual list values?

Reply:
`List[i] -> ScalarList` is a `O(1)` operation in arrow via `list.value(i).into()`, we make it a `O(N)` by casting each element to `scalar`.

Not married, though, was just trying to simplify the `scalar<->array` interaction.

Reply:
I did get partially blocked here because I was unable to use arrow arrays in Ballista's plans, and Ballista does use scalars.

Reply:
Got it, the optimization makes total sense to me. The protobuf ser/der is indeed a tricky problem. I am thinking it's probably better to track this optimization as a separate PR/task to reduce the scope of this PR. There are some other complications introduced from recent changes in datafusions as well including newly added `Hash` and `PartialOrd` traits implementation to the ScalarValue type.

Reply:
For the protobuf ser/der, one thing we could do is to send the `dyan Array` over the wire as arrow ipc stream encoded as bytes in protobuf.

Reply:
Filed https://github.com/apache/arrow-datafusion/issues/996 to track the optimization work as a separate task.